src/rdp.c

```diff
@@ -802,6 +823,169 @@ void rdp_draw_filled_rectangle( int tx, int ty, int bx, int by )
     __rdp_ringbuffer_send();
 }
 
+
+/**
+ * @brief Draw a shaded triangle (no alpha support)
+ *
+ * Draws a shaded triangle to the screen. Vertex order is not important.
+ * Before calling this function, make sure that the RDP is set to shaded mode by
+ * calling #rdp_enable_shading.
+ * Based on triangle guide by snacchus on discord: 
+ * https://docs.google.com/document/d/17ddEo61V0suXbSkKP5mY97QxgUnB-QfAjuBIsPiLWko/edit
+ *
+ * Color values are from 0-255.
+ *
+ * @param[in] x1
+ *            Pixel X1 location of triangle
+ * @param[in] y1
+ *            Pixel Y1 location of triangle
+ * @param[in] x2
+ *            Pixel X2 location of triangle
+ * @param[in] y2
+ *            Pixel Y2 location of triangle
+ * @param[in] x3
+ *            Pixel X3 location of triangle
+ * @param[in] y3
+ *            Pixel Y3 location of triangle
+ * @param[in] v1R
+ *            Red value for vertex 1
+ * @param[in] v1G
+ *            Green value for vertex 1
+ * @param[in] v1B
+ *            Blue value for vertex 1
+ * @param[in] v2R
+ *            Red value for vertex 2
+ * @param[in] v2G
+ *            Green value for vertex 2
+ * @param[in] v2B
+ *            Blue value for vertex 2
+ * @param[in] v3R
+ *            Red value for vertex 3
+ * @param[in] v3G
+ *            Green value for vertex 3
+ * @param[in] v3B
+ *            Blue value for vertex 3
+ */
+void rdp_draw_shaded_triangle(float x1, float y1, float x2, float y2, float x3, float y3, float v1R, float v1G, float v1B, 
+                float v2R, float v2G, float v2B, float v3R, float v3G, float v3B)
+{
+    const float to_fixed_11_2 = 4.0f;
+    const float to_fixed_16_16 = 65536.0f;
+
+    if( y1 > y2 ) { SWAP(y1, y2) SWAP(x1, x2) SWAP(v1R, v2R) SWAP(v1G, v2G) SWAP(v1B, v2B) }
```

Comment:
I think these macro invocations looking like function calls without semicolons are a bit ugly.
You should probably change the SWAP macro to `#define SWAP(a, b) do { float t = a; a = b; b = t; } while (0)`, this would make the macro behave a bit more like normal statements. Also, for consistency, you could add the macro to the existing triangle draw functions as well.

Reply:
Done, thank you.